x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,211 @@
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight := strategy[reqName] // not checking if key is found because it's verified in init()
+
+		for _, score := range scores {
+			newScoreComp := req.Score(*score.Provider)
+			if newScoreComp == 0 {
+				err := fmt.Errorf("score component is zero. score component name: %s, provider address: %s", reqName, score.Provider.Address)
+				panic(err)
+			}
+			newScoreComp = uint64(math.Pow(float64(newScoreComp), float64(weight)))
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// PrepareHashData prepares the hash needed in the pseudo-random choice of providers
+func PrepareHashData(projectIndex string, chainID string, epochHash []byte) []byte {
+	hashData := []byte{}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	return hashData
+}
+
+// PickProviders pick a <group-count> providers set with a pseudo-random weighted choice (using the providers' score list and hashData)
+func PickProviders(ctx sdk.Context, scores []*PairingScore, groupCount int, hashData []byte, chosenProvidersIdx map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	if len(scores) == 0 {
+		return returnedProviders
+	}
+
+	var scoreSum uint64
+	for _, providerScore := range scores {
+		scoreSum += providerScore.Score
+	}
+	if scoreSum == 0 {
+		err := fmt.Errorf("score sum is zero. Cannot pick providers for pairing")
+		panic(err)
+	}
+
+	for it := 0; it < groupCount; it++ {
+		hash := tendermintcrypto.Sha256(hashData) // TODO: we use cheaper algo for speed
+		bigIntNum := new(big.Int).SetBytes(hash)
+		hashAsNumber := sdk.NewUintFromBigInt(bigIntNum)
+		scoreSumUint := sdk.NewUint(scoreSum)
+		modRes := hashAsNumber.Mod(scoreSumUint).Uint64()
+
+		newScoreSum := uint64(0)
+
+		for idx := len(scores) - 1; idx >= 0; idx-- {
+			if chosenProvidersIdx[idx] {
+				// skip index of providers already selected
+				continue
+			}
+			providerScore := scores[idx]
+			newScoreSum += providerScore.Score
+			if modRes < newScoreSum {
+				// we hit our chosen provider
+				returnedProviders = append(returnedProviders, *providerScore.Provider)
+				scoreSum -= providerScore.Score // we remove this provider from the random pool, so the sum is lower now
+				chosenProvidersIdx[idx] = true
+				break
+			}
+		}
+
+		if scoreSum == 0 {
```

Comment:
I still think this is redundant:  IIUC, `scoreSum` will be zero when groupCount is exactly the number of providers _not yet_ selected, and in that case the loop will terminate anyhow.

Reply:
in the case there are less providers than maxProvidersToPair, this scoreSum can be zero before ending the loop

Reply:
oh, right. 
maybe this can be made explicit by iterating on the min(groupCount, len(providers) - len(chosenProviderIdx)) instead?  checking for zero is an indirect test and therefore subtle and more prone to errors should something change later.